### ISSUE TITLE
[java] imports/UnusedImport rule not considering static inner classes of imports

### ISSUE THREAD

**Issue by SUPERCILEX** [pmd/pmd#348]
**Rule Set:** java-imports/UnusedImports

**Code Sample demonstrating the issue:**

```java
import com.supercilex.robotscouter.util.IoHelper;

public class Foo {
    /** {@link SpreadsheetExporter#writeAndShareTeams(Fragment, IoHelper.RequestHandler, List)} */ This should be considered an import
    public void foo() {}
}
```

**Running PMD through:** Gradle

**Comment by jsotuyod** [comment]
I strongly support this one. Until not long ago we had no access to comments from PMD's source. We now do, but we currently don't support Javadoc parsing, which would be necessary to make sure the reference is valid / at a valid point (such as `{@link}` or `{@see}`)

**Comment by SUPERCILEX** [comment]
@jsotuyod well darn it then, I guess this won't be an easy fix. This issue would be a good place to start for javadoc parsing though, so hopefully someone will be motivated to give it a shot.

**Comment by ryan-gustafson** [comment]
@jsotuyod  Comment text has been available for many years.  They are associated to the AST nodes now I take it?  It used to be just a non-contextual list on the compilation unit.

A simple initial approach would be to used substring/regex matching within the content of a tag.  With appropriate pre/post delimiters, it should be reasonably accurate.

**Comment by adangel** [comment]
@ryan-gustafson The comments are first available at the top-level node CompilationUnit. The rules about comments (CommentRequired, ...) map the comments to specific AST nodes.
However, this is not used for the UnusedImport rule.

Here, we just search via regex all FormalComments for any Types. And we actually look already at the parameters of "@link" and "@see" tags. Because of this, I'll change it to a bug...

In this code example, we recognize the parameter "IoHelper.RequestHandler" as a type, but we obviously do not connect it to "IoHelper".

For this rule, there is also a typeresolution aware implementation.

@SUPERCILEX Could you try to use the rule "typeresolution.xml/UnusedImports"? Does this rule report IoHelper as unused import?

**Comment by jsotuyod** [comment]
@ryan-gustafson you are right, I had forgotten about that map. I was thinking of taking them from the token stream directly, but for this issue the map should be enough.

**Comment by SUPERCILEX** [comment]
@adangel The `typeresolution` rule doesn't consider `IoHelper.RequestHandler` to be an unused import! 😄

**Comment by jsotuyod** [comment]
Wow, didn't know the rule was already parsing javadocs. You learn something new everyday.

I'm closing this one then. Typeresolution rules are to be promoted once full type resolution is implemented.

**Comment by SUPERCILEX** [comment]
@jsotuyod Wait, I'm confused. So the `imports.html#UnusedImports` rule is going to be deprecated?

**Comment by jsotuyod** [comment]
@SUPERCILEX it won't. It will be replaced by the implementation currently under typeresolution. The typeresolution ruleset will disappear. Rules from all rulesets may use typeresolution freely.

The typeresolution ruleset was a temporary solution while the typeresolution feature matured.

**Comment by ryan-gustafson** [comment]
Temporary being over a decade. :rofl:

**Comment by SUPERCILEX** [comment]
@ryan-gustafson Ouch.

**Comment by SUPERCILEX** [comment]
@jsotuyod I have some bad news: the reason `typeresolution.html#UnusedImports` ignored my import was because I had forgotten to remove the `NOPMD` comment. So I guess this issue can be opened again, sorry! 😬

**Comment by jsotuyod** [comment]
@SUPERCILEX lol! no problem. Should still be easier to fix than originally expected now I saw the rule is applying regexp on the javadoc.

**Comment by ryan-gustafson** [comment]
@SUPERCILEX Does the actual `@link` text have the 3 asterisks around the `IoHelper.RequestHandler` type, or is that just for emphasis for this issue?

**Comment by SUPERCILEX** [comment]
@ryan-gustafson No, no, no, that was just for extra emphasis. I've removed them from the original comment to reduce confusion. 🎉

**Comment by Fernal73** [comment]
I'm receiving an unused import message while using static imports * in my source.
My source is available at: https://github.com/Fernal73/LearnJava/tree/master/Facade2
I'm using PMD 6. 14.0 with JDK 10 on Arch Linux on Termux.
I don't like importing each static method individually.
Is this problem going away any time soon?
Thank you!

**Comment by jsotuyod** [comment]
@Fernal73 the issue lays in your usage of PMD. You are not using auxclasspath:

https://github.com/Fernal73/LearnJava/blob/9a1196785a72b0a3bd9e212870aa35d3929afb4d/pmd#L17-L20

Without it, PMD can't tell which classes / methods are referenced from which wildcard import. Without auxclasspath PMD falls back to a best effort approach that is limited to the information available on the single source file being looked at a time.

Please refer to the [user docs](https://pmd.github.io/pmd-6.14.0/pmd_userdocs_cli_reference.html) for more info on how to pass along auxclasspath info. Notice that tools such as Gradle and Maven are able to pass along this information for you automatically, but CLI and Ant both need manual configuration.

**Comment by Fernal73** [comment]
Thank you for your prompt response.

Yes, setting the auxclasspath does fix the problem.

However, I now encounter the following issue:

https://github.com/pmd/pmd/issues/1790

The project I'm running it is against is Builder at
https://github.com/Fernal73/LearnJava/tree/master/Builder.

Thank you once again.

Regards,
Linus.

**Comment by Fernal73** [comment]
Thank you very much for your timely response.
Regards,
Linus.

**Comment by Fernal73** [comment]
This is the pmd output file.
Are the classes being imported twice over?

**Comment by Fernal73** [comment]
I'm considering moving to Maven but it's not that easy to customise.
I'm unfamiliar with Gradle.
I'll keep your suggestions in mind.

**Comment by Fernal73** [comment]
I'm receiving a PMD Exception while running pmd against the following file:

package com.javacodegeeks.patterns.singletonpattern;

import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.ObjectInput;
import java.io.ObjectInputStream;
import java.io.ObjectOutput;
import java.io.ObjectOutputStream;
import java.lang.reflect.Constructor;
import java.lang.reflect.Field;
import java.util.Collections;
import java.util.IdentityHashMap;
import java.util.LinkedHashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;
import java.util.concurrent.BrokenBarrierException;
import java.util.concurrent.CyclicBarrier;
import java.util.concurrent.atomic.AtomicReference;

/**
 * Describe class <code>SingletonTest</code> here.
 *
 * @author <a href="mailto:root@localhost"></a>
 * @version 1.0
 */
public class SingletonTest {

  /**
   * Describe <code>main</code> method here.
   *
   * @param args a <code>String</code> value
   */
  public static void main(String[] args) {
    testConcurrency();
    testSerializable();
    testCloneable();
    testReflection();
    testState();
  }

  private static void testConcurrency() {

    final int size = 12;

    final CyclicBarrier cyclicBarrier = new CyclicBarrier(size);

    final AtomicReference<Throwable> exception = new AtomicReference<>();

    // final Set<Long> generatedValues = new HashSet<>(size);
    final Set<Long> generatedValues = new LinkedHashSet<>(size);
    final Set<Singleton> instances =
        Collections.newSetFromMap(new IdentityHashMap<Singleton,
Boolean>());

    final List<Thread> threads = new LinkedList<>();
    for (int i = 0; i < size; i++) {
      final Thread thread = new Thread(new Runnable() {
        @Override
        public void run() {
          try {
            cyclicBarrier.await();
          } catch (InterruptedException | BrokenBarrierException e) {
            exception.compareAndSet(null, e);
            return;
          }

          final Singleton singleton = Singleton.getInstance();
          final long value = singleton.getNextValue();

          // Synchronise the access as the collections used are not
thread-safe
          synchronized (SingletonTest.class) {
            if (!generatedValues.add(value)) {
              exception.compareAndSet(null, new AssertionError("Duplicate
value " + value));
              return;
            }
            instances.add(singleton);
          }
        }
      });
      thread.start();
      threads.add(thread);
    }

    for (final Thread thread : threads) {
      thread.join();
    }

    if (exception.get() != null) {
      throw exception.get();
    }

    switch (instances.size()) {
      case 0:
        throw new AssertionError("Expected one instance, but found none");

      case 1:
        System.out.println("Only one instance created and available.");
        break;
      default:
        throw new AssertionError("Expected one instance, but found many");
    }
    System.out.println("Sequence in order in which inserted: ");
    for (final long value : generatedValues)
      System.out.print(value + " ");
    System.out.println();
  }

  private static void testSerializable() {
    final Singleton instance = Singleton.getInstance();

    final ObjectOutput out = new ObjectOutputStream(new
FileOutputStream("singleton.ser"));
    out.writeObject(instance);
    out.close();

    // deserialize from file to object
    final ObjectInput in = new ObjectInputStream(new
FileInputStream("singleton.ser"));
    final Singleton instance2 = (Singleton) in.readObject();
    in.close();

    System.out.println("instance hashCode:- " + instance.hashCode());
    System.out.println("instance2 hashCode:- " + instance2.hashCode());
  }

  private static void testCloneable() {
    final Singleton obj = Singleton.getInstance();
    try {
      obj.clone();
    } catch (CloneNotSupportedException e) {
      System.out.println(e.getMessage());
    }
  }

  private static void testReflection() {
    try {
      final Constructor<?>[] constructors =
        Singleton.class.getDeclaredConstructors();
      for (Constructor<?> constructor : constructors) {
        constructor.setAccessible(true);
        final Singleton obj = (Singleton) constructor.newInstance();
        System.out.println("obj: Break through Reflection:" + obj);
      }
    } catch (SecurityException se) {
      System.out.println(se.getCause().getMessage());
    }
  }

  private static void testState() {
    resetSingleton();
    final Singleton singleton = Singleton.getInstance();
    if (singleton.getNextValue() != 0L)
      throw new AssertionError("Next value should be zero.");
    resetSingleton();
    singleton = Singleton.getInstance()
    @SuppressWarnings("checkstyle:magicnumber")
    final long expectedValue = 3L;
    singleton.getNextValue();
    singleton.getNextValue();
    singleton.getNextValue();
    if (singleton.getNextValue() != expectedValue)
      throw new AssertionError("Next value should be three.");
    System.out.println("No assert errors. State validated.");
  }

  private static void resetSingleton()
      throws SecurityException,
                      NoSuchFieldException,
                      IllegalArgumentException,
                      IllegalAccessException {
    final Field instance = Singleton.class.getDeclaredField("instance");
    instance.setAccessible(true);
    instance.set(null, null);
  }
}

The file is in the Singleton folder.
I'm using PMD 6.14.0.

Regards,
Linus.

**Comment by Fernal73** [comment]
package com.javacodegeeks.patterns.singletonpattern;

import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.ObjectInput;
import java.io.ObjectInputStream;
import java.io.ObjectOutput;
import java.io.ObjectOutputStream;
import java.lang.reflect.Constructor;
import java.lang.reflect.Field;
import java.lang.reflect.InvocationTargetException;
import java.util.Collections;
import java.util.IdentityHashMap;
import java.util.LinkedHashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;
import java.util.concurrent.BrokenBarrierException;
import java.util.concurrent.CyclicBarrier;
import java.util.concurrent.atomic.AtomicReference;

/**
 * Describe class <code>SingletonTest</code> here.
 *
 * @author <a href="mailto:root@localhost"></a>
 * @version 1.0
 */
public enum SingletonTest {
  ;
  /**
   * Describe <code>main</code> method here.
   *
   * @param args a <code>String</code> value
   */
  public static void main(String[] args) {
    testConcurrency();
    testSerializable();
    testCloneable();
    testReflection();
    testState();
  }

  private static void testConcurrency() {

    final int size = 12;

    final CyclicBarrier cyclicBarrier = new CyclicBarrier(size);

    final AtomicReference<Throwable> exception = new AtomicReference<>();

    // final Set<Long> generatedValues = new HashSet<>(size);
    final Set<Long> generatedValues = new LinkedHashSet<>(size);
    final Set<Singleton> instances =
      Collections.newSetFromMap(new IdentityHashMap<Singleton, Boolean>());

    final List<Thread> threads = new LinkedList<>();
    for (int i = 0; i < size; i++) {
      final Thread thread = new Thread(new Runnable() {
        @Override
        public void run() {
          try {
            cyclicBarrier.await();
          } catch (InterruptedException | BrokenBarrierException e) {
            exception.compareAndSet(null, e);
            return;
          }

          final Singleton singleton = Singleton.getInstance();
          final long value = singleton.getNextValue();

          // Synchronise the access as the collections used are not
thread-safe
          synchronized (SingletonTest.class) {
            if (!generatedValues.add(value)) {
              exception.compareAndSet(null, new AssertionError("Duplicate
value " + value));
              return;
            }
            instances.add(singleton);
          }
        }
      });
      thread.start();
      threads.add(thread);
    }
    try {
      for (final Thread thread : threads)
        thread.join();

      if (exception.get() != null)
        throw exception.get();

      switch (instances.size()) {
        case 0:
          throw new AssertionError("Expected one instance, but found none");

        case 1:
          System.out.println("Only one instance created and available.");
          break;
        default:
          throw new AssertionError("Expected one instance, but found many");
      }
      System.out.println("Sequence in order in which inserted: ");
      for (final long value : generatedValues)
        System.out.print(value + " ");
      System.out.println();
    }
    catch (Throwable throwable) {//NOPMD
      System.out.println(throwable.getMessage());
    }
  }

  private static void testSerializable() {
    final Singleton instance = Singleton.getInstance();

    try {
      final ObjectOutput out = new ObjectOutputStream(new
FileOutputStream("singleton.ser"));
      out.writeObject(instance);
      out.close();

      // deserialize from file to object
      final ObjectInput in = new ObjectInputStream(new
FileInputStream("singleton.ser"));
      final Singleton instance2 = (Singleton) in.readObject();
      in.close();
      System.out.println("instance hashCode:- " + instance.hashCode());
      System.out.println("instance2 hashCode:- " + instance2.hashCode());
    }
    catch (IOException
        | ClassNotFoundException e) {
      System.out.println(e.getMessage());
        }


  }

  private static void testCloneable() {
    final Singleton obj = Singleton.getInstance();
    try {
      obj.clone();
    } catch (CloneNotSupportedException e) {
      System.out.println(e.getMessage());
    }
  }

  private static void testReflection() {
    try {
      final Constructor<?>[] constructors =
        Singleton.class.getDeclaredConstructors();
      for (Constructor<?> constructor : constructors) {
        constructor.setAccessible(true);
        final Singleton obj = (Singleton) constructor.newInstance();
        System.out.println("obj: Break through Reflection:" + obj);
      }
    } catch (SecurityException
        | InstantiationException
        | IllegalArgumentException
        | IllegalAccessException
        | InvocationTargetException e) {
      System.out.println(e.getCause().getMessage());
        }
  }

  private static void testState() {
    try {
      resetSingleton();
      Singleton singleton = Singleton.getInstance();
      if (singleton.getNextValue() != 0L)
        throw new AssertionError("Next value should be zero.");
      resetSingleton();
      singleton = Singleton.getInstance();
      @SuppressWarnings("checkstyle:magicnumber")
      final long expectedValue = 3L;
      singleton.getNextValue();
      singleton.getNextValue();
      singleton.getNextValue();
      if (singleton.getNextValue() != expectedValue)
        throw new AssertionError("Next value should be three.");
      System.out.println("No assert errors. State validated.");
    }
    catch (NoSuchFieldException
        | IllegalAccessException e) {
      System.out.println(e.getMessage());
        }
  }

  private static void resetSingleton()
      throws SecurityException,
                      NoSuchFieldException,
                      IllegalArgumentException,
                      IllegalAccessException {
               final Field instance =
Singleton.class.getDeclaredField("instance");
               instance.setAccessible(true);
               instance.set(null, null);
  }
}

The above is the file updated for compile errors. PMD is probably best run
after compilation unlike Checkstyle.
The PMD Exception does not appear when the code compiles cleanly.

Regards,
Linus.

**Comment by Fernal73** [comment]
An unrelated question but posted here since I'm not sure where to post it
otherwise since you don't seem to have a Google group for PMD.

What is the Maven scope to be used when downloading dependencies from the
Maven repository? Default doesn't give me anything.

<dependency org="net.sourceforge.pmd" name="pmd" rev="6.15.0"
conf="pmd.default-&gt;*">
</dependency>

Is there something I'm missing?

From what I've been researching, there appears to be some way for Ivy to
work with pom files but I'm not quite getting a handle on it as yet.

Your ivy configuration on Maven lists a pom file to be downloaded. Are you
discouraging users from using build environments other than Maven?

Regards,
Linus.

**Comment by jsotuyod** [comment]
@Fernal73 PMD requires code to be syntactically valid to analyze. If it can't compile, it probably can't be analyzed. Moreover, as PMD is best run with the compiled classes in the auxclasspath, it is best to run PMD after compile.

> An unrelated question but posted here since I'm not sure where to post it
otherwise since you don't seem to have a Google group for PMD.

Github issues can be labeled as questions.

> What is the Maven scope to be used when downloading dependencies from the
Maven repository? Default doesn't give me anything.

Maven repository has no idea of scopes. Scopes are only considered locally as to when to include the dependency at all. In Maven itself, it's a plugin dependency, so only used to run the plugin.

I'm unsure of what you are doing with Ivy to run PMD as Ivy only downloads dependencies. If you are using it along with Ant, use whatever conf you want to define as long as that conf is used in Ant to define the PMD task.

> Your ivy configuration on Maven lists a pom file to be downloaded. Are you
discouraging users from using build environments other than Maven?

We don't provide an Ivy configuration… I'm unsure what you are referring to. We do provide support for CLI, Ant, Maven and Gradle, as those are actually able to *run* the PMD task. Again, using Ivy to get dependencies for an Ant build is possible, but completely up to the developer, there is nothing PMD specific there for us to document.

In the future, please consider opening separate issues to keep things tidy / easier to follow.

**Comment by Fernal73** [comment]
https://mvnrepository.com/artifact/net.sourceforge.pmd/pmd/6.15.0

I'm referring to the Ivy configuration listed on the above link.

I've managed to download dependencies for checkstyle and Google java format
using the Ivy configuration below:

<?xml version="1.0"?>

<ivy-module version="2.0">
  <info organisation="linus" module="LearnJava"/>
  <configurations>
    <conf name="checkstyle.main" description="checkstyle main jar"/>
    <conf name="gformat.main" description="google java format main jar"/>
    <conf name="checkstyle.lib" description="checkstyle dependencies"/>
    <conf name="gformat.lib" description="google java format dependencies"/>
    <conf name="pmd.default" description="pmd jars"/>
  </configurations>
  <dependencies>
    
    <dependency org="com.google.googlejavaformat" name="google-java-format"
rev="1.7" conf="gformat.main-&gt;master;gformat.lib-&gt;runtime"/>
    
    <dependency org="com.puppycrawl.tools" name="checkstyle" rev="8.21"
conf="checkstyle.main-&gt;master;checkstyle.lib-&gt;runtime"/>
    
    <dependency org="net.sourceforge.pmd" name="pmd" rev="6.15.0"
conf="pmd.default-&gt;*">
</dependency>
  </dependencies>
</ivy-module>

I'm unable to do something similar for PMD. I'd like to be able to run
PMD as an Ant Task in my project builds under my repository.
But I don't wish to be downloading jars or dependencies I don't need
especially test jars.

What is the best way to do this? I tried 'main' and 'runtime' targets
before but that didn't work.

The repository is at https://github.com/Fernal73/LearnJava

Regards,
Linus.

PS: How do I open new issues and easily? Just email the above github email
address? That works for me.

**Comment by jsotuyod** [comment]
Firstly, that configuration is autogenerated by that site, not ours.

Second, you are looking at the wrong artifact. That is the parent pom for all PMD, not the PMD  modules themselves. You need to use artifacts `pmd-core` and `pmd-java` rather than simply `pmd`. For instance:

```xml
    <dependency org="net.sourceforge.pmd" name="pmd-core" rev="6.15.0"
conf="pmd.default->*">
    <dependency org="net.sourceforge.pmd" name="pmd-java" rev="6.15.0"
conf="pmd.default->*">
```

You may need more language modules if looking to analyze other languages too.

**Comment by Fernal73** [comment]
Thank you very much, Juan.

Much appreciated.

I'll try with these modules and revert if I need any further assistance.

Enjoy the rest of your weekend.

**Comment by Fernal73** [comment]
Now, that I've checked that pmd-core and pmd-java resolve as expected and
the dependencies are downloaded,
I'll try and explain what I meant.

This Stackoverflow post explains how Maven scopes can be mapped to Ivy
configurations.

https://stackoverflow.com/q/7104364/3924108

Unless, my understanding and yours differ, of course.

Secondly, wrt PMD on the Maven repository, there's no documentation that
states that other PMD artifacts are available on the Maven repository.
Perhaps, an organisation page that lists all the artifacts put out by it is
the need of the hour. This should be easily accessible and visible from any
artifact page.

Also, is Github likely to add some sort of Knowledge Base functionality,
that can semantically collage and aggregate user issues and more
importantly resolutions into a living document that can help users find
answers to their questions?

Is there a product that can semantically mark and extract useful snippets
from these Q & A sessions? It does seem a shame if the same queries crop
up, over and over again.

Regards,
Linus.

**Comment by Fernal73** [comment]
collate, not collage.

**Comment by Fernal73** [comment]
I was able to get the PMD Java working.
However, the dependencies as set up in the Ivy. xml and build.xml file
don't download the dependency saxon-9.1. 0.8.jar. The saxon-9.1.0.8-dom.jar
is available though.

The Ivy.xml,classpath file (modified to include saxon-9.1.0.8.jar, not the
generated version), build. xml and pmdjava script are attached.
I
What's the best way to do this?
Wait for a fix, add another dependency to SaxonB or hard-code the jar as
currently done?

Thank you very much,
Linus.

**Comment by adangel** [comment]
I don't know ivy a lot, but maybe you are using only runtime dependencies? With maven, runtime means actually compile+runtime.

I suspect this, because saxon is scope compile and saxon-dom is scope runtime.

If you change this
```xml
<dependency org="net.sourceforge.pmd" name="pmd-core" rev="6.15.0" conf="pmd-core.main-&gt;master;pmd-core.lib-&gt;runtime"></dependency>
```

to

```xml
<dependency org="net.sourceforge.pmd" name="pmd-core" rev="6.15.0" conf="pmd-core.main-&gt;master;pmd-core.lib-&gt;runtime;pmd-core.lib-&gt;compile"></dependency>
```

or simply

```xml
<dependency org="net.sourceforge.pmd" name="pmd-core" rev="6.15.0" conf="pmd-core-&gt;*"></dependency>
```

Do you really need to differ between master artifact and its dependencies?

> Secondly, wrt PMD on the Maven repository, there's no documentation that states that other PMD
> artifacts are available on the Maven repository.

The main (entry) documentation page of PMD is https://pmd.github.io
If you are missing anything, feel free to contribute!

> Also, is Github likely to add some sort of Knowledge Base functionality, that can semantically collage 
> and aggregate user issues and more importantly resolutions into a living document that can help 
> users find answers to their questions?

You'd need to ask github/microsoft for this.
As for our project, we have a section on how to get help: https://pmd.github.io/pmd-6.15.0/pmd_about_help.html
Developers mostly use stackoverflow today for Q&A.

> PS: How do I open new issues and easily? Just email the above github email address? That works 
> for me.

Just use your browser and visit https://github.com/pmd/pmd/issues and hit the "New Issue" button.

> Is there a product that can semantically mark and extract useful snippets from these Q & A sessions? 
> It does seem a shame if the same queries crop up, over and over again.

I'm not aware of such a tool (except for a full text search). The ideal way, how this works, is, that you contribute your findings in a concise documentation, that we can add. This of course requires some effort.

Since you are using Ant as a build too, have you considered using the PMD Ant Task, to execute PMD?
Documentation is here: https://pmd.github.io/pmd-6.15.0/pmd_userdocs_tools_ant.html
Feel free to contribute a similar page describing how to use ivy+ant to download PMD instead of using the binary distribution.

Since you should know now, how to create a new issue, if you have further questions, I'm locking this conversation, because it has nothing at all to do with the UnusedImport rule, this issue is about.
